### PR TITLE
Replace some esp-wifi critical sections with locks

### PR DESCRIPTION
**Pull request by bugadani**
cc #2552, done about half of it. I hope things still work, I didn't test anything.

### CODE REVIEW

File: esp-hal/src/sync.rs [hunk 1]
```diff
@@ -280,6 +280,15 @@ impl<T> Locked<T> {
...
     }
+
+    /// Provide exclusive access to the protected data to the given closure.
+    pub fn with_cs<R>(
+        &self,
+        _cs: critical_section::CriticalSection<'_>,
+        f: impl FnOnce(&T) -> R,
+    ) -> R {
+        f(unsafe { &*self.data.get() })
+    }
```
Comment (Dominaezzz) [inline esp-hal/src/sync.rs:291]: This is not correct. Having a `CriticalSection` doesn't mean you can have access to any other lock.

**Comment by bugadani** [comment]
> So, it definitely makes a difference

I don't know, that difference might just be different optimization choices - there really shouldn't be anything noticeable unless you use multiple cores. Due to reentrancy checks in the new code, I'd be hesitant to accept any perf improvement.

**Comment by bjoernQ** [comment]
> > So, it definitely makes a difference
> 
> I don't know, that difference might just be different optimization choices - there really shouldn't be anything noticeable unless you use multiple cores. Due to reentrancy checks in the new code, I'd be hesitant to accept any perf improvement.

Well, that's the point of my measurement - having lots of interrupts triggered on the second core

**Comment by bugadani** [comment]
Ah, yeah, missed that part 😅 well, that's good then.

**Review by MabezDev — APPROVED** [review]
LGTM, one small nit about an import

File: esp-wifi/src/wifi/mod.rs [hunk 2]
```diff
@@ -2321,26 +2317,22 @@ impl PromiscuousPkt<'_> {
             frame_type,
             len,
             data: core::slice::from_raw_parts(
-                (buf as *const u8).add(size_of::<wifi_pkt_rx_ctrl_t>()),
+                (buf as *const u8).add(core::mem::size_of::<wifi_pkt_rx_ctrl_t>()),
```
[the remark this answers is not in the document]
Comment (bugadani) [reply esp-wifi/src/wifi/mod.rs:2320]: https://www.reddit.com/r/rust/comments/1g6s4ml/psa_size_of_and_align_of_are_in_the_prelude_since/ our msrv is 1.79 so I think it didn't compile?

[the remark this answers is not in the document]
Comment (MabezDev) [reply esp-wifi/src/wifi/mod.rs:2320]: :facepalm: Oops, I think you're right it's just the diff made it look like it was already working in main

[the remark this answers is not in the document]
Comment (bugadani) [reply esp-wifi/src/wifi/mod.rs:2320]: I don't think I've ever compiled feature = sniffer, so I'm not sure if there are more places or not, I also don't remember how exactly I found this.

**Review by bjoernQ — APPROVED** [review]
Thanks

**Pull request merged**